broadcaster/backlog/backlog.go

```diff
@@ -0,0 +1,317 @@
+package backlog
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/log"
+	m "github.com/offchainlabs/nitro/broadcaster/message"
+	"github.com/offchainlabs/nitro/util/arbmath"
+)
+
+var (
+	errDropSegments       = errors.New("remove previous segments from backlog")
+	errSequenceNumberSeen = errors.New("sequence number already present in backlog")
+	errOutOfBounds        = errors.New("message not found in backlog")
+)
+
+type Backlog interface {
+	Head() BacklogSegment
+	Append(bm *m.BroadcastMessage) error
+	Get(start, end uint64) (*m.BroadcastMessage, error)
+	Count() int
+}
+
+type backlog struct {
+	head          atomic.Pointer[backlogSegment]
+	tail          atomic.Pointer[backlogSegment]
+	lookupLock    sync.RWMutex
+	lookupByIndex map[uint64]*atomic.Pointer[backlogSegment]
+	config        ConfigFetcher
+	messageCount  atomic.Uint64
+}
+
+func NewBacklog(c ConfigFetcher) Backlog {
+	lookup := make(map[uint64]*atomic.Pointer[backlogSegment])
+	return &backlog{
+		lookupByIndex: lookup,
+		config:        c,
+	}
+}
+
+func (b *backlog) Head() BacklogSegment {
+	return b.head.Load()
+}
+
+// Append will add the given messages to the backlogSegment at head until
+// that segment reaches its limit. If messages remain to be added a new segment
+// will be created.
+func (b *backlog) Append(bm *m.BroadcastMessage) error {
+
+	if bm.ConfirmedSequenceNumberMessage != nil {
+		b.delete(uint64(bm.ConfirmedSequenceNumberMessage.SequenceNumber))
+		// add to metric?
+	}
+
+	// TODO(clamb): Do I need a max catchup config for the backlog? Similar to catchup buffer
+
+	for _, msg := range bm.Messages {
+		s := b.tail.Load()
+		if s == nil {
+			s = &backlogSegment{}
+			b.head.Store(s)
+			b.tail.Store(s)
+		}
+
+		prevMsgIdx := s.end.Load()
+		if s.count() >= b.config().SegmentLimit {
+			nextS := &backlogSegment{}
+			s.nextSegment.Store(nextS)
+			prevMsgIdx = s.end.Load()
+			nextS.previousSegment.Store(s)
+			s = nextS
+			b.tail.Store(s)
+		}
+
+		err := s.append(prevMsgIdx, msg)
+		if errors.Is(err, errDropSegments) {
+			head := b.head.Load()
+			b.removeFromLookup(head.start.Load(), uint64(msg.SequenceNumber))
+			b.head.Store(s)
+			b.tail.Store(s)
+			b.messageCount.Store(0)
+			log.Warn(err.Error())
+		} else if errors.Is(err, errSequenceNumberSeen) {
+			log.Info("ignoring message sequence number (%s), already in backlog", msg.SequenceNumber)
```

Comment:
We should check that this message is the same as what we have in the buffer. If not, we need to clear the buffer (or preferably only clear this sequence number and anything after it), log an error saying that a reorg occurred, and then add this message.

Reply:
Agreed, the `backlogSegment.append` method has an if statement with the following checks:

1. Is this message the next number in the sequence? +1 from the previous number
2. Else is this message larger than the expected next number in the sequence? If it is return `errDropSegments` error
3. Else this message must have already been seen and we return the `errSequenceNumberSeen` error

Let me know if you think this approach is missing something :)